core/ipc/twin/common/src/main/java/org/opennms/core/ipc/twin/common/AbstractTwinSubscriber.java

```diff
@@ -230,72 +136,183 @@ protected TwinUpdate mapTwinResponseToProto(byte[] responseBytes) {
+         */
+        public synchronized <T> Closeable consume(final Class<T> clazz, final Consumer<T> consumer) {
+            final Consumer<JsonNode> jsonConsumer = (json) -> {
+                try {
+                    // Deserialize to the final class
+                    final var value = AbstractTwinSubscriber.this.objectMapper.treeToValue(json, clazz);
+
+                    // Forward to typed consumer
+                    consumer.accept(value);
+
+                } catch (final JsonProcessingException e) {
+                    LOG.error("Deserialize twin failed: {} as {}", this.key, clazz, e);
+                }
+            };
+
+            if (this.value == null) {
+                // Initially request value
+                this.request();
+            } else {
+                // If value already exists, forward to consumer without requesting
+                jsonConsumer.accept(this.value.value);
+            }
+
+            // Add the consumer to the subscription
+            this.consumers.add(jsonConsumer);
+
+            // Return the closable removing the consumer
+            return () -> {
+                this.consumers.remove(jsonConsumer);
+
+                // Drop the subscription if it becomes empty
+                if (this.consumers.isEmpty()) {
+                    AbstractTwinSubscriber.this.subscriptions.remove(this.key);
+                }
+            };
         }
 
-        @Override
-        public boolean equals(Object o) {
-            if (this == o) return true;
-            if (o == null || getClass() != o.getClass()) return false;
-            SessionImpl<?> session = (SessionImpl<?>) o;
-            return Objects.equals(key, session.key) && Objects.equals(consumer, session.consumer) && Objects.equals(clazz, session.clazz);
+        private synchronized void accept(final Value value) {
+            Objects.requireNonNull(value);
+
+            // Skip if value has not changed
+            if (this.value != null && Objects.equals(this.value.value, value.value)) {
+                return;
+            }
+
+            // Remember value
+            this.value = value;
+
+            // Cancel outstanding retry
+            if (this.retry != null) {
+                this.retry.cancel(false);
+                this.retry = null;
+            }
+
+            // Call all consumers
+            this.consumers.forEach(c -> c.accept(this.value.value));
         }
 
-        @Override
-        public int hashCode() {
-            return Objects.hash(key, consumer, clazz);
+        private synchronized void request() {
+            // Send a request
+            final var request = new TwinRequest(this.key, AbstractTwinSubscriber.this.identity.getLocation());
+            AbstractTwinSubscriber.this.sendRpcRequest(request);
+
+            // Cancel an outstanding retry
+            if (this.retry != null) {
+                this.retry.cancel(false);
```

Comment:
Is there a specific reason why `cancel(false)` is used and not `cancel(true)`?

Reply:
Same as above.